src/robots/msj_platform.cpp

```diff
@@ -157,11 +159,52 @@ class MsjPlatform: public cardsflow::kindyn::Robot{
 
     }
 
+    ///find the closest limit point when the robot is in infeasible state
+    VectorXd findClosestJointLimit(double q0, double q1, double q3){
+        ROS_INFO("FINDING CLOSEST JOINT LIMIT");
+        VectorXd closestLimit;
+        closestLimit.resize(number_of_dofs);
+        closestLimit.setZero();
+        double distance = 0.0, smallestDistance = numeric_limits<double>::max();
+
+        for(int i=0; i < limits[0].size(); i++){
+            VectorXd jointAngle = Vector2d::Zero(), jointLimits = Vector2d::Zero();
+            jointAngle << q0, q1;
+            jointLimits << limits[0][i] ,limits[1][i];
+            distance = (jointAngle - jointLimits).norm();
+            if (distance < smallestDistance){
+                smallestDistance = distance;
+                closestLimit[0] = jointLimits[0];
+                closestLimit[1] = jointLimits[1];
+                closestLimit[2] = q[2];
+            }
+        }
+        return closestLimit;
+    }
+
+    ///set the given joint angle and veloctiy  for each joint
+    void setJointAngleAndVelocity(VectorXd jointAngles, VectorXd jointVel){
+        for(int i=0; i< number_of_dofs; i++){
+            //joint_state[i][0] = jointVel[i];		//Setting velocity to zero send the robot to origin..
+            joint_state[i][1] = jointAngles[i];		//Position of ith joint
+            q[i] = jointAngles[i];
+            //qd[i] = jointVel[i];
+        }
+    }
+
+    ///set the gymstep function response
+    bool setResponse(VectorXd jointAngles,VectorXd jointVel,roboy_simulation_msgs::GymStep::Response &res){
+        for(int i=0; i< number_of_dofs; i++ ){
+            res.q.push_back(jointAngles[i]);
+            res.qdot.push_back(jointVel[i]);
+        }
+    }
+
+
     bool GymStepService(roboy_simulation_msgs::GymStep::Request &req,
                         roboy_simulation_msgs::GymStep::Response &res){
         
         if(req.set_points.size() != 0){ //If no set_point set then just return observation.
-        	//ROS_INFO("Gymstep is called");
         	update();
 	        for(int i=0; i< number_of_cables; i++){
 	        	//Set the commanded tendon velocity from RL agent to simulation 
```

Comment:
Question: Why do we first update, if we have not yet set the new commanded tendon velocity?

Reply:
I don't remember that it's not done in this pr. I could try without but it works now with this.

Reply:
Lets leave it for now